src/L1/rollup/ScrollChain.sol

```diff
@@ -315,78 +347,204 @@ contract ScrollChain is OwnableUpgradeable, PausableUpgradeable, IScrollChain {
-        (bytes32 _parentBatchHash, uint256 _batchIndex, uint256 _totalL1MessagesPoppedOverall) = _beforeCommitBatch(
-            _parentBatchHeader,
-            _chunks
-        );
+        bytes calldata, /*_skippedL1MessageBitmap*/
+        bytes calldata _blobDataProof
+    ) external override {
+        commitBatchWithBlobProof(_version, _parentBatchHeader, _chunks, _blobDataProof);
+    }
 
-        bytes32 _batchHash;
-        uint256 batchPtr;
-        bytes32 _dataHash;
-        uint256 _totalL1MessagesPoppedInBatch;
-        if (1 <= _version && _version <= 2) {
-            // versions 1 and 2 both use ChunkCodecV1 and BatchHeaderV1Codec,
-            // but they use different blob encoding and different verifiers.
-            (_dataHash, _totalL1MessagesPoppedInBatch) = _commitChunksV1(
-                _totalL1MessagesPoppedOverall,
-                _chunks,
-                _skippedL1MessageBitmap
-            );
-            assembly {
-                batchPtr := mload(0x40)
-                _totalL1MessagesPoppedOverall := add(_totalL1MessagesPoppedOverall, _totalL1MessagesPoppedInBatch)
+    /// @inheritdoc IScrollChain
+    function commitBatchWithBlobProof(
+        uint8 _version,
+        bytes calldata _parentBatchHeader,
+        bytes[] memory _chunks,
+        bytes calldata _blobDataProof
+    ) public override OnlySequencer whenEnforcedBatchNotEnable {
+        _commitBatchWithBlobProof(_version, _parentBatchHeader, _chunks, _blobDataProof);
+    }
+
+    /// @inheritdoc IScrollChain
+    function finalizeBundleWithProof(
+        bytes calldata _batchHeader,
+        bytes32 _postStateRoot,
+        bytes32 _withdrawRoot,
+        bytes calldata _aggrProof
+    ) external override OnlyProver whenFinalizeNotPaused {
+        _checkFinalizationAllowed(enforcedBatchParameters.maxFinalizeDelay);
+        _finalizeBundleWithZkProof(_batchHeader, _postStateRoot, _withdrawRoot, _aggrProof);
+    }
+
+    /// @inheritdoc IScrollChain
+    function finalizeBundleWithTeeProof(
+        bytes calldata _batchHeader,
+        bytes32 _postStateRoot,
+        bytes32 _withdrawRoot,
+        bytes calldata _teeProof
+    ) external override OnlyProver whenFinalizeNotPaused {
+        _checkFinalizationAllowed(enforcedBatchParameters.maxFinalizeDelay);
+        _finalizeBundleWithTeeProof(_batchHeader, _postStateRoot, _withdrawRoot, _teeProof);
+    }
+
+    /// @inheritdoc IScrollChain
+    function commitAndFinalizeBatch(CommitStruct calldata commitStruct, FinalizeStruct calldata finalizeStruct)
+        external
+        whenFinalizeNotPaused
+    {
+        //
+        EnforcedBatchParameters memory parameters = enforcedBatchParameters;
+        if (!parameters.enforcedModeEnabled) {
+            uint256 timestamp = IL1MessageQueue(messageQueue).getFirstUnfinalizedMessageTimestamp();
+            if (timestamp > 0 && timestamp + parameters.maxDelayEnterEnforcedMode < block.timestamp) {
```

Comment:
What happens if there's no L1 message (eg because no user posted one on L1)? iiuc then permissionless mode will never be activated. we _could_ make the sending of L1 messages a requirement but so far I thought permissionless mode would be _either_ triggered by a L1 message _or_ just no finalization for too long.

Reply:
oh, yeah you are correct. forget to add this check.